streams/src/main/java/org/apache/kafka/streams/state/internals/InMemoryTimeOrderedKeyValueBuffer.java

```diff
@@ -263,27 +259,28 @@ private void logValue(final Bytes key, final BufferKey bufferKey, final BufferVa
         final ByteBuffer buffer = value.serialize(sizeOfBufferTime);
         buffer.putLong(bufferKey.time());
 
-        collector.send(
-            changelogTopic,
-            key,
-            buffer.array(),
-            V_2_CHANGELOG_HEADERS,
-            partition,
-            null,
-            KEY_SERIALIZER,
-            VALUE_SERIALIZER
+        ((RecordCollector.Supplier) context).recordCollector().send(
```

Comment:
One high-level concern I've developed during this review is whether there's any possibility that something like this could leak into the public API. I.e., is it possible that a Processor, Transformer, or StateStore could have cached some reference from the context that would become invalid when the context gets recycled, similar to the way this recordCollector (which is not public, I know) did.

What's your take on that, @ableegoldman ?

Reply:
Well, only active tasks have a topology, and we don't initialize the topology until everything has been cleanly recycled. So by the time `init` is being called and the context is being used, it should be all up-to-date with the active task references.

Of course that only applies to the Processor/Transformer half of the question. With StateStores we're obviously still calling `init` for standby tasks, and more. But nothing in the public `ProcessorContext` interface gets recycled. Only the cache, record collector, and StreamTask have to be updated, so this should all be totally transparent (unless they're doing something they shouldn't be)

Reply:
Ok. Thanks.